File: src/main/python/AoC2020_24.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 24
#
"""
Coordinate system:
    NW: 1,-1    NE: 0,1
    W : -1,0    0,0         E: 1,0
                SW: 0,-1    SE: 1,-1
"""

from __future__ import annotations

import re
import sys
from collections import Counter
from typing import Iterable
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.game_of_life import GameOfLife


class Direction(NamedTuple):
    q: int
    r: int


DIRS = {
    "ne": Direction(1, -1),
    "nw": Direction(0, -1),
    "se": Direction(0, 1),
    "sw": Direction(-1, 1),
    "e": Direction(1, 0),
    "w": Direction(-1, 0),
}
Tile = tuple[int, int]


class Floor(NamedTuple):
    tiles: set[Tile]

    @classmethod
    def from_input(self, input: list[str]) -> Floor:
        p = re.compile(r"(n?(e|w)|s?(e|w))")
        tiles = set[Tile]()
        for line in input:
            tile = (0, 0)
            for m in p.finditer(line):
                d = DIRS[m.group(0)]
                tile = (tile[0] + d.q, tile[1] + d.r)
            if tile in tiles:
                tiles.remove(tile)
            else:
                tiles.add(tile)
        return Floor(tiles)


class HexGrid(GameOfLife.Universe[Tile]):
    def neighbour_count(self, alive: Iterable[Tile]) -> dict[Tile, int]:
        return Counter(
            (t[0] + d.q, t[1] + d.r) for d in DIRS.values() for t in alive
        )


class Rules(GameOfLife.Rules[Tile]):
    def alive(self, tile: Tile, count: int, alive: Iterable[Tile]) -> bool:
        return count == 2 or (count == 1 and tile in alive)


TEST = """\
sesenwnenenewseeswwswswwnenewsewsw
neeenesenwnwwswnenewnwwsewnenwseswesw
seswneswswsenwwnwse
nwnwneseeswswnenewneswwnewseswneseene
swweswneswnenwsewnwneneseenw
eesenwseswswnenwswnwnwsewwnwsene
sewnenenenesenwsewnenwwwse
wenwwweseeeweswwwnwwe
wsweesenenewnwwnwsenewsenwwsesesenwne
neeswseenwwswnwswswnw
nenwswwsewswnenenewsenwsenwnesesenew
enewnwewneswsewnwswenweswnenwsenwsw
sweneswneswneneenwnewenewwneswswnese
swwesenesewenwneswnwwneseswwne
enesenwswwswneneswsenwnewswseenwsese
wnwnesenesenenwwnenwsewesewsesesew
nenewswnwewswnenesenwnesewesw
eneswnwswnwsenenwnwnwwseeswneewsenese
neswnwewnwnwseenwseesewsenwsweewe
wseweeenwnesenwwwswnew
"""


Input = Floor
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Floor.from_input(list(input_data))

    def part_1(self, floor: Floor) -> int:
        return len(floor.tiles)

    def part_2(self, floor: Floor) -> int:
        gol = GameOfLife(floor.tiles, HexGrid(), Rules())
        for _ in range(100):
            gol.next_generation()
        return sum(1 for _ in gol.alive)

    @aoc_samples(
        (
            ("part_1", TEST, 10),
            ("part_2", TEST, 2208),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2020, 24)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2020_24.java
import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;

import com.github.pareronia.aoc.game_of_life.GameOfLife;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2020_24
            extends SolutionBase<AoC2020_24.Floor, Integer, Integer> {

    private static final Map<String, Direction> DIRS = Map.of(
         "ne", new Direction(1, -1),
         "nw", new Direction(0, -1),
         "se", new Direction(0, 1),
         "sw", new Direction(-1, 1),
         "e", new Direction(1, 0),
         "w", new Direction(-1, 0)
    );
    
	private AoC2020_24(final boolean debug) {
		super(debug);
	}
	
	public static final AoC2020_24 create() {
		return new AoC2020_24(false);
	}

	public static final AoC2020_24 createDebug() {
		return new AoC2020_24(true);
	}
	
	@Override
    protected AoC2020_24.Floor parseInput(final List<String> inputs) {
        return Floor.fromInput(inputs);
    }

    @Override
	public Integer solvePart1(final Floor floor) {
	    return floor.tiles.size();
	}
	
	@Override
	public Integer solvePart2(final Floor floor) {
	    GameOfLife<Tile> gol
	            = new GameOfLife<>(new HexGrid(), new Rules(), floor.tiles);
        for (int i = 0; i < 100; i++) {
            gol = gol.nextGeneration();
        }
        return gol.alive().size();
	}
	
	@Samples({
	    @Sample(method = "part1", input = TEST, expected = "10"),
	    @Sample(method = "part2", input = TEST, expected = "2208"),
	})
	public static void main(final String[] args) throws Exception {
	    AoC2020_24.create().run();
	}
	
	private static final String TEST = """
	        sesenwnenenewseeswwswswwnenewsewsw
	        neeenesenwnwwswnenewnwwsewnenwseswesw
	        seswneswswsenwwnwse
	        nwnwneseeswswnenewneswwnewseswneseene
	        swweswneswnenwsewnwneneseenw
	        eesenwseswswnenwswnwnwsewwnwsene
	        sewnenenenesenwsewnenwwwse
	        wenwwweseeeweswwwnwwe
	        wsweesenenewnwwnwsenewsenwwsesesenwne
	        neeswseenwwswnwswswnw
	        nenwswwsewswnenenewsenwsenwnesesenew
	        enewnwewneswsewnwswenweswnenwsenwsw
	        sweneswneswneneenwnewenewwneswswnese
	        swwesenesewenwneswnwwneseswwne
	        enesenwswwswneneswsenwnewswseenwsese
	        wnwnesenesenenwwnenwsewesewsesesew
	        nenewswnwewswnenesenwnesewesw
	        eneswnwswnwsenenwnwnwwseeswneewsenese
	        neswnwewnwnwseenwseesewsenwsweewe
	        wseweeenwnesenwwwswn
	        """;
	
	record Tile(int q, int r) {
		
		public static Tile at(final int q, final int r) {
		    return new Tile(q, r);
		}
		
		public Tile at(final Direction direction) {
		    return Tile.at(this.q + direction.q, this.r + direction.r);
		}
	}

	record Direction(int q, int r) {}
	
	record Floor(Set<Tile> tiles) {

        public static Floor fromInput(final List<String> input) {
            final Pattern pattern = Pattern.compile("(n?(e|w)|s?(e|w))");
            final Set<Tile> tiles = new HashSet<>();
            for (final String line : input) {
                final Tile tile = pattern.matcher(line).results()
                    .map(MatchResult::group)
                    .map(DIRS::get)
                    .reduce(Tile.at(0, 0), Tile::at, (t1, t2) -> t2);
                if (tiles.contains(tile)) {
                    tiles.remove(tile);
                } else {
                    tiles.add(tile);
                }
            }
            return new Floor(tiles);
        }
	}
	
	private static final class HexGrid implements GameOfLife.Type<Tile> {

        @Override
        public Map<Tile, Long> getNeighbourCounts(final Set<Tile> alive) {
            return alive.stream()
                .flatMap(tile -> DIRS.values().stream().map(tile::at))
                .collect(groupingBy(tile -> tile, counting()));
        }
	}
	
	private static final class Rules implements GameOfLife.Rules<Tile> {

        @Override
        public boolean alive(final Tile cell, final long cnt, final Set<Tile> alive) {
            return cnt == 2 || (cnt == 1 && alive.contains(cell));
        }
	}
}
